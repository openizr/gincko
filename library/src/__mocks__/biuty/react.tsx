/**
 * Copyright (c) Openizr. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import * as React from 'react';

/**
 * biuty/react mock.
 */
function Component(props: JSXElement): JSX.Element {
  const { onChange, onFocus, onClick } = props;
  const { transform, onKeyDown, onBlur } = props;

  // Covers `onChange` handler.
  setTimeout(() => {
    onChange?.('test');
    if (process.env.IS_DATE === 'true') {
      onChange?.('2020/02/20');
    }
  }, 10);

  // Covers `onFocus` handler.
  onFocus?.();

  // Covers `onBlur` handler.
  onBlur?.();

  // Covers `transform` function.
  transform?.('ok');
  transform?.('1002');
  transform?.('100220');
  transform?.('10022020');

  // Covers `onKeyDown` handler.
  onKeyDown?.('', { key: '1' });
  onKeyDown?.('', { key: 'A', ctrlKey: false, preventDefault: vi.fn() });
  // Covers `onClick` handler.
  if (process.env.NOCLICK !== 'true') {
    setTimeout(() => {
      onClick?.();
    }, 10);
  }
  return (
    <div>{JSON.stringify(props)}</div>
  );
}

export const UILink = Component;
export const UIButton = Component;
export const UIOptions = Component;
export const UITextfield = Component;
export const UITextarea = Component;
export const UIFilePicker = Component;
export const buildClass = vi.fn((...values: string[]): string => values.join(' '));
export const markdown = vi.fn((label: string, lightMode: boolean) => `MARKDOWN FOR ${label}, ${lightMode}`);
