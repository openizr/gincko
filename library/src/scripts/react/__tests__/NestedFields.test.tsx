/**
 * Copyright (c) Openizr. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @jest-environment jsdom
 */

import React from 'react';
import NestedFields from 'scripts/react/NestedFields';
import { render, waitFor } from '@testing-library/react';

jest.mock('biuty/react');
jest.mock('scripts/react/Field');

describe('react/NestedFields', () => {
  const onUserAction = jest.fn();
  const userInputs = { test: 'ok' };
  const variables = { var: 'value' };
  const i18n = jest.fn(() => 'TRANSLATED LABEL');

  beforeEach(() => {
    jest.clearAllMocks();
    delete process.env.NOCLICK;
  });

  test('renders correctly - array', async () => {
    const { container } = render(<NestedFields
      t={i18n}
      type="array"
      label="Test"
      helper="Helper"
      path="path.0.to.field"
      variables={variables}
      userInputs={userInputs}
      fields={[{
        id: 'Message',
        status: 'initial',
        component: 'Message',
        componentProps: {},
      }, null]}
      onUserAction={onUserAction}
    />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders correctly - array with minItems', async () => {
    const { container } = render(<NestedFields
      t={i18n}
      type="array"
      minItems={2}
      path="path.0.to.field"
      variables={variables}
      userInputs={userInputs}
      fields={[{
        id: 'Message',
        status: 'initial',
        component: 'Message',
        componentProps: {},
      }]}
      onUserAction={onUserAction}
    />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders correctly - array with maxItems', async () => {
    const { container } = render(<NestedFields
      t={i18n}
      type="array"
      maxItems={1}
      path="path.0.to.field"
      variables={variables}
      userInputs={userInputs}
      value={['test', 'test']}
      fields={[{
        id: 'Message',
        status: 'initial',
        component: 'Message',
        componentProps: {},
      }, null]}
      onUserAction={onUserAction}
    />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders correctly - object', async () => {
    const { container } = render(<NestedFields
      t={i18n}
      type="object"
      path="path.0.to.field"
      variables={variables}
      userInputs={userInputs}
      fields={[{
        id: 'Message',
        status: 'initial',
        component: 'Message',
        componentProps: {},
      }, null]}
      onUserAction={onUserAction}
    />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders correctly - dynamicObject with valid pattern', async () => {
    process.env.NOCLICK = 'true';
    const { container } = render(<NestedFields
      t={i18n}
      type="dynamicObject"
      path="path.0.to.field"
      variables={variables}
      userInputs={userInputs}
      fields={[{
        id: 'Message',
        status: 'initial',
        component: 'Message',
        componentProps: {},
      }, null]}
      onUserAction={onUserAction}
      allowedPatterns={[/test/]}
    />);
    await waitFor(() => new Promise<void>((resolve) => { setTimeout(resolve, 50); }));
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders correctly - dynamicObject with invalid pattern', async () => {
    const { container } = render(<NestedFields
      t={i18n}
      type="dynamicObject"
      path="path.0.to.field"
      variables={variables}
      userInputs={userInputs}
      fields={[{
        id: 'Message',
        status: 'initial',
        component: 'Message',
        componentProps: {},
      }, null]}
      onUserAction={onUserAction}
      addTextfieldProps={{ debounceTimeout: 10, helper: 'INVALID' }}
    />);
    await waitFor(() => new Promise<void>((resolve) => { setTimeout(resolve, 50); }));
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders correctly - dynamicObject with error', async () => {
    const { container } = render(<NestedFields
      t={i18n}
      helper="Error"
      modifiers="error"
      type="dynamicObject"
      variables={variables}
      path="path.0.to.field"
      userInputs={userInputs}
      value={{
        test: 'Test',
        test2: 'Test',
      }}
      fields={[{
        id: 'Message',
        status: 'initial',
        component: 'Message',
        componentProps: {},
      }, {
        id: 'Message2',
        status: 'initial',
        component: 'Message',
        componentProps: {},
      }, null]}
      onUserAction={onUserAction}
    />);
    await waitFor(() => new Promise<void>((resolve) => { setTimeout(resolve, 50); }));
    expect(container.firstChild).toMatchSnapshot();
  });
});
