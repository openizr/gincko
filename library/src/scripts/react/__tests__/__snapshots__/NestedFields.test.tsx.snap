// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`react/NestedFields > renders correctly - array 1`] = `
<div
  class="ui-nested-fields  array"
>
  <span
    class="ui-nested-fields__label"
  >
    Test
  </span>
  <div
    class="ui-nested-fields__field"
  >
    <div>
      {"type":"button"}
    </div>
    <div
      id="field"
    >
      {"field":{"id":"Message","status":"initial","component":"Message","componentProps":{}},"isActive":false,"variables":{"var":"value"},"userInputs":{"test":"ok"},"path":"path.0.to.field.Message","customComponents":{}}
    </div>
  </div>
  <div
    class="ui-nested-fields__add"
  >
    <div>
      {"type":"button","modifiers":" "}
    </div>
  </div>
  <span
    class="ui-nested-fields__helper"
  >
    Helper
  </span>
</div>
`;

exports[`react/NestedFields > renders correctly - array with maxItems 1`] = `
<div
  class="ui-nested-fields  array"
>
  <div
    class="ui-nested-fields__field"
  >
    <div>
      {"type":"button"}
    </div>
    <div
      id="field"
    >
      {"field":{"id":"Message","status":"initial","component":"Message","componentProps":{}},"isActive":false,"variables":{"var":"value"},"userInputs":{"test":"ok"},"path":"path.0.to.field.Message","customComponents":{}}
    </div>
  </div>
  <div
    class="ui-nested-fields__add"
  />
</div>
`;

exports[`react/NestedFields > renders correctly - array with minItems 1`] = `
<div
  class="ui-nested-fields  array"
>
  <div
    class="ui-nested-fields__field"
  >
    <div
      id="field"
    >
      {"field":{"id":"Message","status":"initial","component":"Message","componentProps":{}},"isActive":false,"variables":{"var":"value"},"userInputs":{"test":"ok"},"path":"path.0.to.field.Message","customComponents":{}}
    </div>
  </div>
  <div
    class="ui-nested-fields__add"
  >
    <div>
      {"type":"button","modifiers":" "}
    </div>
  </div>
</div>
`;

exports[`react/NestedFields > renders correctly - dynamicObject with error 1`] = `
<div
  class="ui-nested-fields error dynamicObject"
>
  <div
    class="ui-nested-fields__field"
  >
    <span
      class="ui-nested-fields__field__label"
    >
      Message
    </span>
    <div>
      {"type":"button"}
    </div>
    <div
      id="field"
    >
      {"field":{"id":"Message","status":"initial","component":"Message","componentProps":{}},"isActive":false,"variables":{"var":"value"},"userInputs":{"test":"ok"},"path":"path.0.to.field.Message","customComponents":{}}
    </div>
  </div>
  <div
    class="ui-nested-fields__field"
  >
    <span
      class="ui-nested-fields__field__label"
    >
      Message2
    </span>
    <div>
      {"type":"button"}
    </div>
    <div
      id="field"
    >
      {"field":{"id":"Message2","status":"initial","component":"Message","componentProps":{}},"isActive":false,"variables":{"var":"value"},"userInputs":{"test":"ok"},"path":"path.0.to.field.Message2","customComponents":{}}
    </div>
  </div>
  <div
    class="ui-nested-fields__add"
  >
    <div>
      {"value":"","name":"undefined-add","readonly":true,"debounceTimeout":100}
    </div>
    <div>
      {"type":"button","modifiers":"disabled "}
    </div>
  </div>
  <span
    class="ui-nested-fields__helper"
  >
    Error
  </span>
</div>
`;

exports[`react/NestedFields > renders correctly - dynamicObject with invalid pattern 1`] = `
<div
  class="ui-nested-fields  dynamicObject"
>
  <div
    class="ui-nested-fields__field"
  >
    <span
      class="ui-nested-fields__field__label"
    >
      Message
    </span>
    <div>
      {"type":"button"}
    </div>
    <div
      id="field"
    >
      {"field":{"id":"Message","status":"initial","component":"Message","componentProps":{}},"isActive":false,"variables":{"var":"value"},"userInputs":{"test":"ok"},"path":"path.0.to.field.Message","customComponents":{}}
    </div>
  </div>
  <div
    class="ui-nested-fields__add"
  >
    <div>
      {"value":"","name":"undefined-add","readonly":true,"helper":"INVALID","debounceTimeout":10}
    </div>
    <div>
      {"type":"button","modifiers":"disabled "}
    </div>
  </div>
</div>
`;

exports[`react/NestedFields > renders correctly - dynamicObject with valid pattern 1`] = `
<div
  class="ui-nested-fields  dynamicObject"
>
  <div
    class="ui-nested-fields__field"
  >
    <span
      class="ui-nested-fields__field__label"
    >
      Message
    </span>
    <div>
      {"type":"button"}
    </div>
    <div
      id="field"
    >
      {"field":{"id":"Message","status":"initial","component":"Message","componentProps":{}},"isActive":false,"variables":{"var":"value"},"userInputs":{"test":"ok"},"path":"path.0.to.field.Message","customComponents":{}}
    </div>
  </div>
  <div
    class="ui-nested-fields__add"
  >
    <div>
      {"value":"test","name":"undefined-add","readonly":true,"debounceTimeout":100}
    </div>
    <div>
      {"type":"button","modifiers":" "}
    </div>
  </div>
</div>
`;

exports[`react/NestedFields > renders correctly - object 1`] = `
<div
  class="ui-nested-fields  object"
>
  <div
    class="ui-nested-fields__field"
  >
    <div
      id="field"
    >
      {"field":{"id":"Message","status":"initial","component":"Message","componentProps":{}},"isActive":false,"variables":{"var":"value"},"userInputs":{"test":"ok"},"path":"path.0.to.field.Message","customComponents":{}}
    </div>
  </div>
</div>
`;
